### DISCUSSION TITLE
How to setup propietary PDK with .yml file

### DISCUSSION THREAD

**Discussion by Z-KN** [ucb-bar/hammer#855] [Support (Q&A)] [answered]
I am going to use the hammer flow; but encountered the following problem.

1. I want to bind the technology as our own PDK, TSMC PDK. But the tutorial [https://chipyard.readthedocs.io/en/latest/VLSI/Hammer.html#technology-plugins](https://urldefense.com/v3/__https://chipyard.readthedocs.io/en/latest/VLSI/Hammer.html*technology-plugins__;Iw!!DZ3fjg!8IIxK1INMWFFD2u0bk8hQFpR_4BDgtacQNIeA33e6Y9zDTN09UZ2wjRbelCiRB1j9S3RgGc8nzkuZoKgVhkQ2jUXHKEpSq0$) only describes the setting up process for proprietary PDK vaguely. I wonder how to write the `.yml` file to link to our own PDK.

1. Next, should I run something like  `git submodule add git@my-secure-server.edu:xxx/hammer-tsmc-plugin.git` under the `vlsi` directory? If yes,  what should I put inside the hammer-tsmc-plugin repository before running the previous command?

1. I tried to run `.script/vlsi-init.sh tsmc`, but it reports: `Failed to clone 'vlsi/hammer-mentor-plugins'`. Retry scheduled I have the access to mentor on our server. How to resolve this problem?

**Comment by harrisonliew** [comment]
Hi,

1. Please refer to Hammer's documentation, https://hammer-vlsi.readthedocs.io/en/latest/Technology/index.html. Section 2.2 Hammer Tech JSON describes how to specify all the library files, cells, and stackups necessary to describe the PDK.
2. Yes, you will need to add the git submodule that way. Section 2.1 of that documentation describes what required files there need to be in the plugin.
3. I have added you to the hammer-mentor-plugins repository now. You should be able to clone it now.

By the way, there is an open pull request with updates to the documentation for how to write the tech JSON construct using Pydantic models (maybe more intuitive than the JSON file). If you wish to do that instead, please refer to #848.

**Reply by Z-KN** [reply]
Thank you for that! So do you mean I have to write a `.tech.json` file like this https://github.com/ucb-bar/hammer/blob/master/hammer/technology/asap7/asap7.tech.json, for my own PDK?
And I should put something whose structure is similar to https://github.com/ucb-bar/hammer/tree/master/hammer/technology/asap7 inside `hammer-tsmc-plugin` dir?

**Reply by harrisonliew** [reply, marked as answer]
Yes, and yes. Note that the directory structure inside your plugin will be like:
```
hammer-tsmc-plugin
-> hammer
    -> tsmc
        -> <files>
```
And then when you select the plugin, you will specify it as `vlsi.core.technology: hammer.tsmc`.

**Reply by Z-KN** [reply]
Thank you very much!
